# How to analyze HUMAnN2 outputs

__From Yurgel et al. 2019 (Frontier in Microbiology)__:

1.HUMAnN2 0.11.2 (Franzosa et al., 2018) was run on these filtered reads to identify the number of __reads per kilobase (RPK)__ of UniRef50 gene families (Suzek et al., 2015) in each sample. UniRef50 gene families were then converted into KEGG orthologs using HUMAnN2’s built in mapping files. 
2.KEGG ortholog abundances were then _normalized by the number of genome equivalents_ (average genome size/library size) found in each sample using MicrobeCensus (Nayfach and Pollard, 2015) to get __reads per kilobase per genome equivalent (RPKG)__. 
3.Normalized KEGG orthologs were then mapped to KEGG pathways and KEGG modules using MinPath (Ye and Doak, 2009) and the PICRUSt2 script pathway_pipeline.py (Douglas et al., 2019) using the “–no_regroup” option and mapping files included with PICRUSt2 v2.1.4-b. MetaPhlAn2 v2.7.62 (Truong et al., 2015) was run with default options within HUMAnN2 to identify the relative abundances of taxa within each sample. 
4.A custom Python script was used to determine the GC-content of the microbiome overall and for the V. corymbosum reference genome.
5.HUMAnN2 tables with RPKG of metabolic pathways were transformed to BIOM tables and used for the generation of beta-diversity (weighted Bray–Curtis dissimilarity) (Bray and Curtis, 1957) metrics and analysis of variations in sample groupings explained by weighted Bray–Curtis beta-diversity dissimilarity (Adonis tests, 999 permutations) using QIIME wrapper scripts (Caporaso et al., 2010). An important assumption of this approach is that the read depth per sample is sufficiently high to accurately estimate the abundance of the gene families underlying these metabolic pathways. Analysis of taxonomic and functional profiles was performed using the STAMP software package (Parks et al., 2014). Corrected p-values (q-values) were calculated based on Benjamini–Hochberg FDR multiple-test correction. Features with (Welch’s t-test) q-value < 0.05 were considered significant and were thus retained.